src/osdc/Filer.h

```diff
@@ -43,12 +45,74 @@ class OSDMap;
 class Finisher;
 
 
+struct C_TruncRange;
+
+class TruncWorkQueue : public md_config_obs_t {
+  CephContext *cct;
+  Objecter *objecter;
+  Finisher *finisher;
+  Throttle throttle;
+  std::queue<C_TruncRange*> queue;
+  std::thread thread;
+  std::mutex lock;
+  std::condition_variable cv;
+  bool stopping;
+  using lock_guard = std::lock_guard<std::mutex>;
+  using unique_lock = std::unique_lock<std::mutex>;
+
+
+  void run();
+  void update_throttle_maximum();
+  C_TruncRange* dequeue() {
+    C_TruncRange *c_trunc_range = nullptr;
+    {
+      unique_lock lk(lock);
+	    cv.wait(lk, [this]{return stopping || !queue.empty() ;});
+      if (!stopping) {
+        c_trunc_range = queue.front();
+        queue.pop();
+	    }
+    }
+    return c_trunc_range;
+  }
+  bool queue_job_to_finisher(C_TruncRange* c_tr);
+
+public:
+  TruncWorkQueue(Objecter *o, Finisher *f)
+      : cct(o->cct), objecter(o), finisher(f),
+        throttle(o->cct, "trunc_op_throttle",
+                 o->cct->_conf.get_val<uint64_t>("osdc_max_truncate_ops")),
+        thread(&TruncWorkQueue::run, this), stopping(false) {
+    cct->_conf.add_observer(this);
+  }
+  ~TruncWorkQueue() {
+    cct->_conf.remove_observer(this);
+  };
+  void set_stopping() {
+    {
+      lock_guard lk(lock);
+      stopping = true;
+      cv.notify_one();
+    }
+    thread.join();
+  }
+  bool is_stopping() { return stopping; }
+  void enqueue(C_TruncRange *ctr);
+  void put_slots(int put) { throttle.put(put); };
+
+  /// md_config_obs_t
+  const char** get_tracked_conf_keys() const override;
+  void handle_conf_change(const ConfigProxy& conf,
+			                    const std::set<std::string> &changed) override;
+};
+
 /**** Filer interface ***/
 
 class Filer {
   CephContext *cct;
   Objecter   *objecter;
   Finisher   *finisher;
+  TruncWorkQueue *trunc_work_queue;
```

Comment:
Why not make this part of the `Filer`? Then mds changes would not be required.

Reply:
Only `MDCache` use `Filer::truncate()` and it means only `MDCache` need `TruncWorkQueue`.
The other classes will create many unnecessary work queues (with thread) if I make `TruncWorkQueue` of the `Filer`.
Even if I use constructor overloading, I still need to control the timing of stopping `TruncWorkQueue`.

Reply:
Should I use constructor overloading to avoid unnecessary work queues and add `Filer::shutdown()` ?
It still changes mds (`trunc_work_queue.set_stopping();` -> `filer.shutdown()`) but may more makes sense.

Reply:
I think you can use a `std::unique_ptr` for the TruncWorkQueue in Filer, then check if it's not instantiated before use in `Filer::_do_truncate_range`. Finally, yes, add a `Filer::is_active`  method which the MDS calls in `MDCache::shutdown_pass`. We don't want to shutdown the truncate work queue while some files are being purged. Instead, the MDS should wait until finished.

When the Filer is destructed, I think it should be good time to call the shutdown method on the truncate work queue, no?

Reply:
I don't understand why `Filer::is_active` should be added because the truncate work queue can do nothing with stopped finisher.
PS.
`MDCache::shutdown_pass` already check the status of the objecter of mds. (Line 8016 ~ 8021)
The objecter is shut down after the finisher is stopped in `MDSRankDispatcher::shutdown()`.

It's all the truncate work queue can do to stop send contexts to finisher and cancel all contexts in the queue (when the finisher is stopped).
Stop send contexts to finisher:
`trunc_work_queue.set_stopping()`
Wait for cancel all contexts in the queue: (After setting stopping)
`thread.join()`

So I think the three parts should be fixed:
1. Filer will create the truncate work queue in accordance with the new parameter `bool create_trunc_work_queue`.
2. Move the code `thread.join();` from `TruncWorkQueue::set_stopping()` to `~TruncWorkQueue()`.
3. Add `Filer::set_stopping()` which will call `TruncWorkQueue::set_stopping()` and let `MDCache::shutdown()` call it.